cleanlab/datalab/internal/issue_manager/underperf_group.py

```diff
@@ -126,24 +127,84 @@ def set_knn_graph(self, features: npt.NDArray, find_issues_kwargs: Dict) -> csr_
             except NotFittedError:
                 knn.fit(features)
 
-            knn_graph = knn.kneighbors_graph(mode="distance", n_neighbors=features.shape[0] - 1)
-        return knn_graph
+            knn_graph = knn.kneighbors_graph(mode="distance")
+            symmetric_knn_graph = self._symmetrize_knn_graph(knn_graph)
+        return symmetric_knn_graph, knn_graph
+
+    def _symmetrize_knn_graph(self, knn_graph: csr_matrix) -> csr_matrix:
+        binary_adjacency = (knn_graph > 0).astype(int)
+        symmetric_adjacency = binary_adjacency.maximum(binary_adjacency.T)
+        symmetric_distances = knn_graph + knn_graph.T.multiply(
+            symmetric_adjacency - binary_adjacency
+        )
+        return symmetric_distances
+
+    def perform_clustering(self, knn_graph: csr_matrix) -> npt.NDArray[np.int_]:
+        # hdb = HDBSCAN(metric="precomputed")
+        # hdb.fit(knn_graph)
+        # cluster_labels = hdb.labels_
+        # ALTERNATIVE FORMULATION: Connect the graph using a dummy node before HDBSCAN.
+        # Adapted from - https://github.com/scikit-learn-contrib/hdbscan/issues/82#issuecomment-1242604639
+        # knn_graph = self.connect_knn_graph(knn_graph)
+        cluster_labels = self.run_hdbscan_on_components(knn_graph)
+
+        return cluster_labels[:200]  # Slicing applicable if connect_knn_graph is called
+
+    def run_hdbscan_on_components(
+        self, knn_graph: csr_matrix, min_cluster_size: int = 5, **hdbscan_params: Dict[str, Any]
+    ) -> npt.NDArray[np.int_]:
+        n_components, component_labels = sp.csgraph.connected_components(
+            csgraph=knn_graph, directed=False, return_labels=True
+        )
+        all_clusters = np.empty(knn_graph.shape[0], dtype=object)
+        offset = 0
+        for component_idx in range(n_components):
+            component_mask = component_labels == component_idx
+            component_subgraph = knn_graph[component_mask, :][:, component_mask]
+            knn_subgraph = self._symmetrize_knn_graph(component_subgraph)
+            clusterer = HDBSCAN(metric="precomputed")
+            component_clusters = clusterer.fit_predict(knn_subgraph)
+            outliers_mask = np.in1d(component_clusters, self.OUTLIER_LABELS)
+            component_clusters[~outliers_mask] += offset
+            all_clusters[component_mask] = component_clusters
+            offset = np.max(component_clusters) + 1
+        return all_clusters
```

Comment:
Good to keep this. I'd suggest you make this an "internal" method. If we extract _symmetrize_knn_graph into a function, then this could be a function as well that we move to a separate utility module.

For now, just rename it to `_run_hdbscan_on_components`.